jina/drivers/__init__.py

```diff
@@ -159,7 +159,76 @@ def __getstate__(self):
         return d
 
 
-class BaseExecutableDriver(BaseDriver):
+class BaseRecursiveDriver(BaseDriver):
+
+    def __init__(self, depth_range: Tuple[int] = (0, 0), order: str = 'post', *args, **kwargs):
+        """
+
+        :param depth_range: right-exclusive range of the recursion depth, (0,0) for root-level only
+        :param order: the traverse and apply order. if 'post' then first traverse then call apply, if 'pre' then first apply then traverse
+        :param args:
+        :param kwargs:
+        """
+        super().__init__(*args, **kwargs)
+        self._depth_start = depth_range[0]
+        self._depth_end = depth_range[1]
+        if order in {'post', 'pre'}:
+            self.recursion_order = order
+        else:
+            raise AttributeError('can only accept oder={"pre", "post"}')
+
+    def apply(self, doc: 'jina_pb2.Document', *args, **kwargs):
```

Comment:
I think we should add the @abc.abstractmethod decorator

Reply:
I might be wrong, not every Driver implements apply, am I right?

Reply:
ControlDriver doesn't need, only the Drivers that operate on `docs` need to implement `apply` or `apply_all` or both.

Reply:
ok!